src/Data/Text/Markdown.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Data.Text.Markdown where
+
+import CMarkGFM qualified as CMark
+import Control.Applicative ((<|>))
+import Control.Monad (void)
+import Data.Function ((&))
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Yaml
+import Swarm.Language.Module (moduleAST)
+import Swarm.Language.Parse (readTerm)
+import Swarm.Language.Pipeline (ProcessedTerm (..), prettyTypeErr, processParsedTerm)
+import Swarm.Language.Pretty (prettyText)
+import Swarm.Language.Syntax (Syntax)
+
+data Node
+  = Node NodeType [Node]
+  | LeafText Text
+  | LeafCode Syntax
+  | LeafCodeBlock String Syntax
+  deriving (Eq, Show)
+
+data NodeType
+  = Document
+  | Paragraph
+  deriving (Eq, Show)
+
+instance ToJSON Node where
+  toJSON = String . toText
+
+instance FromJSON Node where
+  parseJSON v =
+    withArray "markdown paragraphs" (foldMap parseJSON) v
+      <|> withText "markdown text" fromTextM v
+instance Semigroup Node where
+  Node Document cs1 <> Node Document cs2 = Node Document (cs1 <> cs2)
+  Node Document cs1 <> n2 = Node Document (cs1 <> [n2])
+  n1 <> Node Document cs2 = Node Document (n1 : cs2)
+  n1 <> n2 = Node Document [n1, n2]
+
+instance Monoid Node where
+  mempty = Node Document []
```

Comment:
This is not actually an identity element.  For example `LeafText "t" <> mempty = Node Document [LeafText "t"] /= LeafText "t"`.  I'm not sure how much this matters.  I suppose it is an identity if we consider `Node Document [x] = x`.

Reply:
I think all parsed markdown will have document as the root node.

Maybe I won’t need the instance in the end, but it could be handy for preserving the paragraph lists syntax in scenarios.

Reply:
The types should make more sense now. 🙂 There is still some magic around nesting, but it stays at two levels only (document and paragraph).